In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline

np.set_printoptions(threshold=20, precision=2, suppress=True)
pd.options.display.max_rows = 7
pd.options.display.max_columns = 8
pd.set_option('precision', 2)
# This option stops scientific notation for pandas
# pd.set_option('display.float_format', '{:.2f}'.format)


## Part 1: Getting Started


In [2]:
pd.DataFrame?

In [3]:
df_rand = pd.DataFrame(np.random.randint(low=0, high=100, size=(5, 5)),
                       columns=['a', 'b', 'c', 'd', 'e'])
df_rand

a   b   c   d   e
0  33  22  60  42  55
1  44  60  48  50  63
2  17  44  27  56  10
3  87  40  23  38  22
4  40  61  77  55  30

In [4]:
df_rand = pd.DataFrame(np.random.randint(1, 10, (5, 5)),
                       columns=[1, 2, 3, 4, 5])
df_rand

1  2  3  4  5
0  9  8  4  4  8
1  7  9  9  9  3
2  5  4  7  4  7
3  6  1  3  5  2
4  2  8  8  5  9

In [5]:
df_rand.iloc[-1]

1    2
2    8
3    8
4    5
5    9
Name: 4, dtype: int64

In [6]:
# example adopted from 
# https://stackoverflow.com/questions/31593201/how-are-iloc-ix-and-loc-different
 
df = pd.DataFrame({'Age': [30, 20, 22, 40, 32, 28, 39],
                   'Color': ['Blue', 'Green', 'Red', 'White', 'Gray', 'Black','Red'],
                   'Food': ['Pasta', 'Fish', 'Mango', 'Apple', 'Cheese','Melon', 'Beans'],
                   'Height': [165, 70, 120, 80, 180, 172, 150],
                   'Score': [4.6, 8.3, 9.0, 3.3, 1.8, 9.5, 2.2],
                   'State': ['NY', 'HI', 'FL', 'AL', 'AK', 'TX', 'CA']
                   },
                  index=['Jane', 'Kahea', 'Aaron', 'Penelope', 'Dean', 'Christina', 'Cornelia'])
 
df

Age  Color    Food  Height  Score State
Jane        30   Blue   Pasta     165    4.6    NY
Kahea       20  Green    Fish      70    8.3    HI
Aaron       22    Red   Mango     120    9.0    FL
Penelope    40  White   Apple      80    3.3    AL
Dean        32   Gray  Cheese     180    1.8    AK
Christina   28  Black   Melon     172    9.5    TX
Cornelia    39    Red   Beans     150    2.2    CA

In [7]:
df.shape

(7, 6)

In [8]:
df.count()

Age       7
Color     7
Food      7
Height    7
Score     7
State     7
dtype: int64

In [9]:
df.dtypes

Age         int64
Color      object
Food       object
Height      int64
Score     float64
State      object
dtype: object

In [10]:
df_obj = pd.DataFrame(data=df, dtype=np.object)

In [11]:
df_obj.dtypes

Age       object
Color     object
Food      object
Height    object
Score     object
State     object
dtype: object

In [12]:
df_obj['Age'] < 30

Jane         False
Kahea         True
Aaron         True
Penelope     False
Dean         False
Christina     True
Cornelia     False
Name: Age, dtype: bool

In [13]:
(df['Age'] < 30).values

array([False,  True,  True, False, False,  True, False])

In [14]:
print("\n -- loc -- \n")
print(df.loc[df['Age'] < 30, ['Color', 'Height']])

print("\n -- iloc -- \n")
print(df.iloc[(df['Age'] < 30).values, [1, 3]])


 -- loc -- 

           Color  Height
Kahea      Green      70
Aaron        Red     120
Christina  Black     172

 -- iloc -- 

           Color  Height
Kahea      Green      70
Aaron        Red     120
Christina  Black     172


In [15]:
df.loc['Cornelia']

Age          39
Color       Red
Food      Beans
Height      150
Score       2.2
State        CA
Name: Cornelia, dtype: object

In [16]:
df.loc['Kahea':'Dean']

Age  Color    Food  Height  Score State
Kahea      20  Green    Fish      70    8.3    HI
Aaron      22    Red   Mango     120    9.0    FL
Penelope   40  White   Apple      80    3.3    AL
Dean       32   Gray  Cheese     180    1.8    AK

In [17]:
# Transpose the table
df.T

Jane  Kahea  Aaron Penelope    Dean Christina Cornelia
Age        30     20     22       40      32        28       39
Color    Blue  Green    Red    White    Gray     Black      Red
Food    Pasta   Fish  Mango    Apple  Cheese     Melon    Beans
Height    165     70    120       80     180       172      150
Score     4.6    8.3      9      3.3     1.8       9.5      2.2
State      NY     HI     FL       AL      AK        TX       CA

In [18]:
dft = df.T
dft.loc['Food']

Jane          Pasta
Kahea          Fish
Aaron         Mango
Penelope      Apple
Dean         Cheese
Christina     Melon
Cornelia      Beans
Name: Food, dtype: object

Now we can read in the data using `pd.read_csv` ([docs](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html)).

In [19]:
baby = pd.read_csv('babynames.csv')
baby

Name Sex  Count  Year
0            Mary   F   9217  1884
1            Anna   F   3860  1884
2            Emma   F   2587  1884
...           ...  ..    ...   ...
1891891     Verna   M      5  1883
1891892    Winnie   M      5  1883
1891893  Winthrop   M      5  1883

[1891894 rows x 4 columns]

Note that for the code above to work, the `babynames.csv` file must be located in the same directory as this notebook. We can check what files are in the current folder by running the `ls` command-line tool (you can also omit the `!`, since it's one of the most-used commands; for more information, see [Chapter 1: IPython and Shell Commands](https://jakevdp.github.io/PythonDataScienceHandbook/01.05-ipython-and-shell-commands.html)):

In [20]:
!ls

Emp.pickle                            df_nob.pickle
apply-vs-agg-Copy1.ipynb              lec03-pandas.ipynb
apply-vs-agg-Copy2.ipynb              lec04-pandas.ipynb
apply-vs-agg.ipynb                    missing-values-correlation-problem.py
babynames.csv                         mtcars.pickle
df.pickle                             pandas_apply_strings_plotting.ipynb
df_b.pickle


When we use `pandas` to read in data, we get a DataFrame. A DataFrame is a tabular data structure where each column is labeled (in this case 'Name', 'Sex', 'Count', 'Year') and each row is labeled (in this case 0, 1, 2, ..., 1891893).

The labels of a DataFrame are called the *indexes* of the DataFrame and make many data manipulations easier.

## Indexes, Slicing, and Sorting

Let's use `pandas` to answer the following question:

**What were the five most popular baby names in 2016?**

### Breaking the Problem Down

We can decompose this question into the following simpler table manipulations:

1. Slice out the rows for the year 2016.
2. Sort the rows in descending order by Count.

Now, we can express these steps in `pandas`.

### Slicing using `.loc`

To select subsets of a DataFrame, we use the `.loc` slicing syntax. The first argument is the label of the row and the second is the label of the column:

In [21]:
baby

Name Sex  Count  Year
0            Mary   F   9217  1884
1            Anna   F   3860  1884
2            Emma   F   2587  1884
...           ...  ..    ...   ...
1891891     Verna   M      5  1883
1891892    Winnie   M      5  1883
1891893  Winthrop   M      5  1883

[1891894 rows x 4 columns]

In [22]:
baby.loc[0, 'Name'] # Row labeled 1, Column labeled 'Name'

'Mary'

To slice out multiple rows or columns, we can use `:`. Note that **`.loc` slicing is inclusive, unlike Python's slicing**.

In [23]:
# Get rows 1 through 5, columns Name through Count inclusive
baby.loc[1:5, 'Name':'Count']

Name Sex  Count
1       Anna   F   3860
2       Emma   F   2587
3  Elizabeth   F   2549
4     Minnie   F   2243
5   Margaret   F   2142

### Slight tangent: exploring slicing, indexing and manipulation of DataFrames and Series

We will often want a single column from a DataFrame:

In [24]:
baby.loc[:, 'Year']

0          1884
1          1884
2          1884
           ... 
1891891    1883
1891892    1883
1891893    1883
Name: Year, Length: 1891894, dtype: int64

Note that when we select a single column, we get a `pandas` Series. A Series is like a one-dimensional NumPy array since we can perform arithmetic on all the elements at once.

In [25]:
baby.loc[:, 'Year'] * 2

0          3768
1          3768
2          3768
           ... 
1891891    3766
1891892    3766
1891893    3766
Name: Year, Length: 1891894, dtype: int64

To select out specific columns, we can pass a list into the `.loc` slice:

In [26]:
# This is a DataFrame again
baby.loc[:, ['Name', 'Year']]

Name  Year
0            Mary  1884
1            Anna  1884
2            Emma  1884
...           ...   ...
1891891     Verna  1883
1891892    Winnie  1883
1891893  Winthrop  1883

[1891894 rows x 2 columns]

Selecting columns is common, so there's a shorthand.

In [27]:
# Shorthand for baby.loc[:, 'Name']
baby['Name']

0              Mary
1              Anna
2              Emma
             ...   
1891891       Verna
1891892      Winnie
1891893    Winthrop
Name: Name, Length: 1891894, dtype: object

In [28]:
# Shorthand for baby.loc[:, ['Name', 'Count']]
baby[['Name', 'Count']]

Name  Count
0            Mary   9217
1            Anna   3860
2            Emma   2587
...           ...    ...
1891891     Verna      5
1891892    Winnie      5
1891893  Winthrop      5

[1891894 rows x 2 columns]

OK, end of tangent.

Let's get back to the original task.

#### Slicing rows using a predicate

To slice out the rows with year 2016, we will first create a Series containing `True` for each row we want to keep and `False` for each row we want to drop. This is simple because math and boolean operators on Series are applied to each element in the Series.

In [29]:
# Series of years
baby['Year']

0          1884
1          1884
2          1884
           ... 
1891891    1883
1891892    1883
1891893    1883
Name: Year, Length: 1891894, dtype: int64

In [30]:
# Compare each year with 2016
baby['Year'] == 2016

0          False
1          False
2          False
           ...  
1891891    False
1891892    False
1891893    False
Name: Year, Length: 1891894, dtype: bool

Once we have this Series of `True` and `False`, we can pass it into `.loc`.

In [31]:
# We are slicing rows, so the boolean Series goes in the first
# argument to .loc
baby_2016 = baby.loc[baby['Year'] == 2016, :]
baby_2016

Name Sex  Count  Year
1850880    Emma   F  19414  2016
1850881  Olivia   F  19246  2016
1850882     Ava   F  16237  2016
...         ...  ..    ...   ...
1883745  Zyahir   M      5  2016
1883746    Zyel   M      5  2016
1883747   Zylyn   M      5  2016

[32868 rows x 4 columns]

### Sorting Rows

The next step is the sort the rows in descending order by 'Count'. We can use the `sort_values()` function.

In [32]:
sorted_2016 = baby_2016.sort_values('Count', ascending=False)
sorted_2016

Name Sex  Count  Year
1850880      Emma   F  19414  2016
1850881    Olivia   F  19246  2016
1869637      Noah   M  19015  2016
...           ...  ..    ...   ...
1868752  Mikaelyn   F      5  2016
1868751    Miette   F      5  2016
1883747     Zylyn   M      5  2016

[32868 rows x 4 columns]

Finally, we will use `.iloc` to slice out the first five rows of the DataFrame. **`.iloc` works like `.loc` but takes in numerical indices instead of labels**. It **does not include the right endpoint** in its slices, like Python's list slicing.

In [33]:
# Get the value in the zeroth row, zeroth column
sorted_2016.iloc[0, 0]

'Emma'

In [34]:
# Get the first five rows
sorted_2016.iloc[0:5]

Name Sex  Count  Year
1850880    Emma   F  19414  2016
1850881  Olivia   F  19246  2016
1869637    Noah   M  19015  2016
1869638    Liam   M  18138  2016
1850882     Ava   F  16237  2016

## In Conclusion

We now have the five most popular baby names in 2016 and learned to express the following operations in `pandas`:

| Operation | `pandas` |
| --------- | -------  |
| Read a CSV file | `pd.read_csv()` |
| Slicing using labels or indices | `.loc` and `.iloc` |
| Slicing rows using a predicate | Use a boolean-valued Series in `.loc` |
| Sorting rows | `.sort_values()` |

## Part 2: Grouping and Pivoting

In this section, we will answer the question:

**What were the most popular male and female names in each year?**

Here's the Baby Names dataset once again:

In [35]:
baby = pd.read_csv('babynames.csv')
baby.head()
# the .head() method outputs the first five rows of the DataFrame

Name Sex  Count  Year
0       Mary   F   9217  1884
1       Anna   F   3860  1884
2       Emma   F   2587  1884
3  Elizabeth   F   2549  1884
4     Minnie   F   2243  1884

### Breaking the Problem Down

We should first notice that the question in the previous section has similarities to this one; the question in the previous section restricts names to babies born in 2016 whereas this question asks for names in all years.

We once again decompose this problem into simpler table manipulations.

1. Group the `baby` DataFrame by 'Year' and 'Sex'.
2. For each group, compute the most popular name.

Recognizing which operation is needed for each problem is sometimes tricky. Usually, a convoluted series of steps will signal to you that there might be a simpler way to express what you want. If we didn't immediately recognize that we needed to group, for example, we might write steps like the following:

1. Loop through each unique year.
2. For each year, loop through each unique sex.
3. For each unique year and sex, find the most common name.

There is almost always a better alternative to looping over a `pandas` DataFrame. **In particular, looping over unique values of a DataFrame should usually be replaced with a group.**

### Grouping

To group in `pandas`. we use the `.groupby()` method.

In [36]:
baby.groupby('Year')

`.groupby()` returns a strange-looking `DataFrameGroupBy` object. We can call `.agg()` on this object with an aggregation function in order to get a familiar output:

In [37]:
# The aggregation function takes in a series of values for each group
# and outputs a single value
def length(series):
    return len(series)

# Count up number of values for each year. This is equivalent to
# counting the number of rows where each year appears.
baby.groupby('Year').agg(length)

Name    Sex  Count
Year                     
1880   2000   2000   2000
1881   1935   1935   1935
1882   2127   2127   2127
...     ...    ...    ...
2014  33206  33206  33206
2015  33063  33063  33063
2016  32868  32868  32868

[137 rows x 3 columns]

You might notice that the `length` function simply calls the `len` function, so we can simplify the code above.

In [38]:
baby.groupby('Year').agg(len)

Name    Sex  Count
Year                     
1880   2000   2000   2000
1881   1935   1935   1935
1882   2127   2127   2127
...     ...    ...    ...
2014  33206  33206  33206
2015  33063  33063  33063
2016  32868  32868  32868

[137 rows x 3 columns]

The aggregation is applied to each column of the DataFrame, producing redundant information. We can restrict the output columns by slicing before grouping.

In [39]:
year_rows = baby[['Year', 'Count']].groupby('Year').agg(len)
year_rows

# A further shorthand to accomplish the same result:
#
# year_counts = baby[['Year', 'Count']].groupby('Year').count()
#
# pandas has shorthands for common aggregation functions, including
# count, sum, and mean.

Count
Year       
1880   2000
1881   1935
1882   2127
...     ...
2014  33206
2015  33063
2016  32868

[137 rows x 1 columns]

Note that the index of the resulting DataFrame now contains the unique years, so we can slice subsets of years using `.loc` as before:

In [40]:
# Every twentieth year starting at 1880
year_rows.loc[1880:2016:20, :]

Count
Year       
1880   2000
1900   3730
1920  10755
1940   8961
1960  11924
1980  19440
2000  29764

### Grouping on Multiple Columns

We can group on multiple columns to get groups based on unique pairs of values. To do this, pass in a list of column labels into `.groupby()`.

In [41]:
grouped_counts = baby.groupby(['Year', 'Sex']).max()
grouped_counts

Name  Count
Year Sex              
1880 F     Zula   7065
     M     Zeke   9655
1881 F     Zula   6919
...         ...    ...
2015 M    Zyvon  19594
2016 F     Zyva  19414
     M    Zyrus  19015

[274 rows x 2 columns]

The code above computes the total number of babies born for each year and sex. Let's now use grouping by muliple columns to compute the most popular names for each year and sex. Since the data are already sorted in descending order of Count for each year and sex, we can define an aggregation function that returns the first value in each series. (If the data weren't sorted, we can call `sort_values()` first.)

In [42]:
# The most popular name is simply the first one that appears in the series
def most_popular(series):
    return series.iloc[0]

baby_pop = baby.groupby(['Year', 'Sex']).agg(most_popular)
baby_pop

Name  Count
Year Sex             
1880 F    Mary   7065
     M    John   9655
1881 F    Mary   6919
...        ...    ...
2015 M    Noah  19594
2016 F    Emma  19414
     M    Noah  19015

[274 rows x 2 columns]

Notice that grouping by multiple columns results in multiple labels for each row. This is called a "multilevel index" and is tricky to work with. The important thing to know is that **`.loc` takes in a tuple for the row index instead of a single value**:

In [43]:
baby_pop.loc[(2000, 'M'), 'Name']

'Jacob'

But `.iloc` behaves the same as usual since it uses indices instead of labels:

In [44]:
baby_pop.iloc[10:15, :]

Name  Count
Year Sex             
1885 F    Mary   9128
     M    John   8756
1886 F    Mary   9889
     M    John   9026
1887 F    Mary   9888

### Pivoting

**If you group by two columns, you can often use pivot to present your data in a more convenient format.** Using a pivot lets you use one set of grouped labels as the columns of the resulting table.

To pivot, use the `pd.pivot_table()` function.

In [45]:
pd.pivot_table(baby,
               index='Year',         # Index for rows
               columns='Sex',        # Columns
               values='Name',        # Values in table
               aggfunc=most_popular) # Aggregation function

Sex      F     M
Year            
1880  Mary  John
1881  Mary  John
1882  Mary  John
...    ...   ...
2014  Emma  Noah
2015  Emma  Noah
2016  Emma  Noah

[137 rows x 2 columns]

Compare this result to the `baby_pop` table that we computed using `.groupby()`. We can see that the `Sex` index in `baby_pop` became the columns of the pivot table.

In [46]:
baby_pop

Name  Count
Year Sex             
1880 F    Mary   7065
     M    John   9655
1881 F    Mary   6919
...        ...    ...
2015 M    Noah  19594
2016 F    Emma  19414
     M    Noah  19015

[274 rows x 2 columns]

## In Conclusion

We now have the most popular baby names for each sex and year in our dataset and learned to express the following operations in `pandas`:

| Operation | `pandas` |
| --------- | -------  |
| Group | `df.groupby(label)` |
| Group by multiple columns | `df.groupby([label1, label2])` |
| Group and aggregate | `df.groupby(label).agg(func)` |
| Pivot | `pd.pivot_table()` |